File: src/translate.rs
use serde_json::{json, Value};

use crate::{session::SessionStore, types::*};

/// Convert a Responses API request + prior history into a Chat Completions request.
pub fn to_chat_request(req: &ResponsesRequest, history: Vec<ChatMessage>, sessions: &SessionStore) -> ChatRequest {
    let mut messages = history;

    // Prefer `instructions` (Codex CLI) over `system` (other clients).
    let system_text = req.instructions.as_ref().or(req.system.as_ref());
    if let Some(system) = system_text {
        if messages.is_empty() || messages[0].role != "system" {
            messages.insert(
                0,
                ChatMessage {
                    role: "system".into(),
                    content: Some(system.clone()),
                    reasoning_content: None,
                    tool_calls: None,
                    tool_call_id: None,
                    name: None,
                },
            );
        }
    }

    // Append new input, mapping Responses API roles to Chat Completions roles.
    match &req.input {
        ResponsesInput::Text(text) => {
            messages.push(ChatMessage {
                role: "user".into(),
                content: Some(text.clone()),
                reasoning_content: None,
                tool_calls: None,
                tool_call_id: None,
                name: None,
            });
        }
        ResponsesInput::Messages(items) => {
            // Process items with index so we can group consecutive function_call
            // entries into a single assistant message. Providers require all tool
            // calls from one turn to live in one message with a tool_calls array.
            let mut i = 0;
            while i < items.len() {
                let item = &items[i];
                let item_type = item.get("type").and_then(|v| v.as_str()).unwrap_or("");

                if item_type == "function_call" {
                    // Collect this and all immediately following function_call items
                    // into one assistant message with multiple tool_calls entries.
                    let mut grouped: Vec<Value> = Vec::new();
                    let mut reasoning_content: Option<String> = None;

                    while i < items.len() {
                        let cur = &items[i];
                        if cur.get("type").and_then(|v| v.as_str()).unwrap_or("") != "function_call" {
                            break;
                        }
                        let call_id = cur.get("call_id").and_then(|v| v.as_str()).unwrap_or("");
                        let name    = cur.get("name").and_then(|v| v.as_str()).unwrap_or("");
                        let args    = cur.get("arguments").and_then(|v| v.as_str()).unwrap_or("{}");
                        if reasoning_content.is_none() {
                            reasoning_content = sessions.get_reasoning(call_id);
                        }
                        grouped.push(json!({
                            "id": call_id,
                            "type": "function",
                            "function": { "name": name, "arguments": args }
                        }));
                        i += 1;
                    }

                    let mut msg = ChatMessage {
                        role: "assistant".into(),
                        content: None,
                        reasoning_content,
                        tool_calls: Some(grouped),
                        tool_call_id: None,
                        name: None,
                    };
                    // Fallback: try turn-level fingerprint if call_id lookup missed
                    if msg.reasoning_content.is_none() {
                        msg.reasoning_content = sessions.get_turn_reasoning(&messages, &msg);
                    }
                    messages.push(msg);
                } else {
                    match item_type {
                        "function_call_output" => {
                            let call_id = item.get("call_id").and_then(|v| v.as_str()).unwrap_or("");
                            let output  = item.get("output").and_then(|v| v.as_str()).unwrap_or("");
                            messages.push(ChatMessage {
                                role: "tool".into(),
                                content: Some(output.to_string()),
                                reasoning_content: None,
                                tool_calls: None,
                                tool_call_id: Some(call_id.to_string()),
                                name: None,
                            });
                        }
                        _ => {
                            // Regular user/assistant/developer message
                            let role = item.get("role").and_then(|v| v.as_str()).unwrap_or("user");
                            let role = match role {
                                "developer" => "system",
                                other => other,
                            }
                            .to_string();
                            let content = value_to_text(item.get("content"));
                            let mut msg = ChatMessage {
                                role,
                                content: Some(content),
                                reasoning_content: None,
                                tool_calls: None,
                                tool_call_id: None,
                                name: None,
                            };
                            // For assistant messages, try to recover reasoning_content
                            // from the turn-level index (needed for thinking models like
                            // DeepSeek that require reasoning_content to be passed back).
                            if msg.role == "assistant" {
                                msg.reasoning_content = sessions.get_turn_reasoning(&messages, &msg);
                            }
                            messages.push(msg);
                        }
                    }
                    i += 1;
                }
            }
        }
    }

    ChatRequest {
        model: req.model.clone(),
        messages,
        // Keep only `function` tools; providers like DeepSeek don't accept
        // OpenAI-proprietary built-ins (web_search, computer, file_search, …).
        tools: req.tools.iter()
            .filter(|t| t.get("type").and_then(Value::as_str) == Some("function"))
            .map(convert_tool)
            .collect(),
        temperature: req.temperature,
        max_tokens: req.max_output_tokens,
        stream: req.stream,
    }
}

/// Responses API tool format → Chat Completions tool format.
///
/// Responses API (flat):
///   {"type":"function","name":"foo","description":"...","parameters":{...},"strict":false}
///
/// Chat Completions (nested):
///   {"type":"function","function":{"name":"foo","description":"...","parameters":{...}}}
fn convert_tool(tool: &Value) -> Value {
    let Some(obj) = tool.as_object() else {
        return tool.clone();
    };
    // Already in Chat Completions format if it has a "function" sub-object.
    if obj.contains_key("function") {
        return tool.clone();
    }
    // Convert from Responses API flat format.
    if obj.get("type").and_then(Value::as_str) == Some("function") {
        let mut func = serde_json::Map::new();
        if let Some(v) = obj.get("name") { func.insert("name".into(), v.clone()); }
        if let Some(v) = obj.get("description") { func.insert("description".into(), v.clone()); }
        if let Some(v) = obj.get("parameters") { func.insert("parameters".into(), v.clone()); }
        if let Some(v) = obj.get("strict") { func.insert("strict".into(), v.clone()); }
        return json!({"type": "function", "function": func});
    }
    tool.clone()
}

/// Convert a Chat Completions response into a Responses API response.
pub fn from_chat_response(
    id: String,
    model: &str,
    chat: ChatResponse,
) -> (ResponsesResponse, Vec<ChatMessage>) {
    let choice = chat.choices.into_iter().next().unwrap_or_else(|| ChatChoice {
        message: ChatMessage {
            role: "assistant".into(),
            content: Some(String::new()),
            reasoning_content: None,
            tool_calls: None,
            tool_call_id: None,
            name: None,
        },
    });

    let text = choice.message.content.clone().unwrap_or_default();
    let usage = chat.usage.unwrap_or(ChatUsage {
        prompt_tokens: 0,
        completion_tokens: 0,
        total_tokens: 0,
    });

    let response = ResponsesResponse {
        id,
        object: "response",
        model: model.to_string(),
        output: vec![ResponsesOutputItem {
            kind: "message".into(),
            role: "assistant".into(),
            content: vec![ContentPart {
                kind: "output_text".into(),
                text: Some(text),
            }],
        }],
        usage: ResponsesUsage {
            input_tokens: usage.prompt_tokens,
            output_tokens: usage.completion_tokens,
            total_tokens: usage.total_tokens,
        },
    };

    (response, vec![choice.message])
}

/// Collapse a Responses API content value (string or parts array) to plain text.
fn value_to_text(v: Option<&Value>) -> String {
    match v {
        None => String::new(),
        Some(Value::String(s)) => s.clone(),
        Some(Value::Array(parts)) => parts
            .iter()
            .filter_map(|p| p.get("text").and_then(|t| t.as_str()))
            .collect::<Vec<_>>()
            .join(""),
        Some(other) => other.to_string(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn base_req(input: ResponsesInput) -> ResponsesRequest {
        ResponsesRequest {
            model: "test".into(),
            input,
            previous_response_id: None,
            tools: vec![],
            stream: false,
            temperature: None,
            max_output_tokens: None,
            system: None,
            instructions: None,
        }
    }

    #[test]
    fn test_text_input_becomes_user_message() {
        let sessions = SessionStore::new();
        let req = base_req(ResponsesInput::Text("hello".into()));
        let chat = to_chat_request(&req, vec![], &sessions);
        assert_eq!(chat.messages.len(), 1);
        assert_eq!(chat.messages[0].role, "user");
        assert_eq!(chat.messages[0].content.as_deref(), Some("hello"));
    }

    #[test]
    fn test_system_prompt_from_instructions() {
        let sessions = SessionStore::new();
        let mut req = base_req(ResponsesInput::Text("hi".into()));
        req.instructions = Some("be helpful".into());
        let chat = to_chat_request(&req, vec![], &sessions);
        assert_eq!(chat.messages[0].role, "system");
        assert_eq!(chat.messages[0].content.as_deref(), Some("be helpful"));
    }

    #[test]
    fn test_developer_role_mapped_to_system() {
        let sessions = SessionStore::new();
        let req = base_req(ResponsesInput::Messages(vec![
            json!({"type": "message", "role": "developer", "content": "secret instructions"}),
        ]));
        let chat = to_chat_request(&req, vec![], &sessions);
        assert_eq!(chat.messages[0].role, "system");
        assert_eq!(chat.messages[0].content.as_deref(), Some("secret instructions"));
    }

    #[test]
    fn test_function_call_grouping() {
        let sessions = SessionStore::new();
        let req = base_req(ResponsesInput::Messages(vec![
            json!({"type": "function_call", "call_id": "c1", "name": "fn_a", "arguments": "{}"}),
            json!({"type": "function_call", "call_id": "c2", "name": "fn_b", "arguments": "{}"}),
        ]));
        let chat = to_chat_request(&req, vec![], &sessions);
        assert_eq!(chat.messages.len(), 1);
        assert_eq!(chat.messages[0].role, "assistant");
        let calls = chat.messages[0].tool_calls.as_ref().unwrap();
        assert_eq!(calls.len(), 2);
        assert_eq!(calls[0]["id"], "c1");
        assert_eq!(calls[1]["id"], "c2");
    }

    #[test]
    fn test_function_call_output_becomes_tool_message() {
        let sessions = SessionStore::new();
        let req = base_req(ResponsesInput::Messages(vec![
            json!({"type": "function_call_output", "call_id": "c1", "output": "result"}),
        ]));
        let chat = to_chat_request(&req, vec![], &sessions);
        assert_eq!(chat.messages[0].role, "tool");
        assert_eq!(chat.messages[0].content.as_deref(), Some("result"));
        assert_eq!(chat.messages[0].tool_call_id.as_deref(), Some("c1"));
    }

    #[test]
    fn test_convert_tool_flat_to_nested() {
        let flat = json!({
            "type": "function",
            "name": "my_fn",
            "description": "does stuff",
            "parameters": {"type": "object"}
        });
        let nested = convert_tool(&flat);
        assert_eq!(nested["type"], "function");
        assert_eq!(nested["function"]["name"], "my_fn");
        assert_eq!(nested["function"]["description"], "does stuff");
    }

    #[test]
    fn test_convert_tool_already_nested() {
        let already = json!({
            "type": "function",
            "function": {"name": "my_fn", "description": "does stuff"}
        });
        let result = convert_tool(&already);
        assert_eq!(result, already);
    }

    #[test]
    fn test_value_to_text_string() {
        let sessions = SessionStore::new();
        let req = base_req(ResponsesInput::Messages(vec![
            json!({"type": "message", "role": "user", "content": "plain text"}),
        ]));
        let chat = to_chat_request(&req, vec![], &sessions);
        assert_eq!(chat.messages[0].content.as_deref(), Some("plain text"));
    }

    #[test]
    fn test_value_to_text_parts_array() {
        let sessions = SessionStore::new();
        let req = base_req(ResponsesInput::Messages(vec![
            json!({"type": "message", "role": "user", "content": [
                {"type": "input_text", "text": "hello "},
                {"type": "input_text", "text": "world"}
            ]}),
        ]));
        let chat = to_chat_request(&req, vec![], &sessions);
        assert_eq!(chat.messages[0].content.as_deref(), Some("hello world"));
    }
}

File: translate.ts
import type {
  ResponsesRequest,
  ResponsesResponse,
  ResponsesOutputItem,
  ResponsesUsage,
  ContentPart,
  ChatRequest,
  ChatMessage,
  ChatResponse,
  ChatChoice,
  ChatUsage,
} from "./types.ts";
import { SessionStore } from "./config.ts";
import { createModuleLogger } from "./logger.ts";

const log = createModuleLogger("translate");

/// Convert a Responses API request + prior history into a Chat Completions request.
export function toChatRequest(
  req: ResponsesRequest,
  history: ChatMessage[],
  sessions: SessionStore
): ChatRequest {
  log.debug("开始转换请求", { model: req.model, history_len: history.length });

  const messages: ChatMessage[] = [...history];

  // Prefer `instructions` (Codex CLI) over `system` (other clients).
  const systemText = req.instructions || req.system;
  if (systemText) {
    if (messages.length === 0 || messages[0].role !== "system") {
      messages.unshift({
        role: "system",
        content: systemText,
      });
      log.debug("添加 system message", { content: systemText });
    }
  }

  // Append new input, mapping Responses API roles to Chat Completions roles.
  const input = req.input;
  if (typeof input === "string") {
    // Plain text input
    messages.push({
      role: "user",
      content: input,
    });
    log.debug("文本输入转为 user message", { content: input });
  } else if (Array.isArray(input)) {
    // Process items array
    log.debug("处理 input 数组", { items_count: input.length });

    let i = 0;
    while (i < input.length) {
      const item = input[i] as Record<string, unknown>;
      const itemType = (item?.type as string) || "";

      if (itemType === "function_call") {
        // Collect consecutive function_call items into one assistant message
        const grouped: Record<string, unknown>[] = [];
        let reasoningContent: string | undefined = undefined;

        while (i < input.length) {
          const cur = input[i] as Record<string, unknown>;
          if ((cur?.type as string) !== "function_call") break;

          const callId = (cur?.call_id as string) || "";
          const name = (cur?.name as string) || "";
          const args = (cur?.arguments as string) || "{}";

          if (!reasoningContent) {
            reasoningContent = sessions.getReasoning(callId);
          }

          grouped.push({
            id: callId,
            type: "function",
            function: { name, arguments: args },
          });
          i++;
        }

        const msg: ChatMessage = {
          role: "assistant",
          content: undefined,
          reasoning_content: reasoningContent,
          tool_calls: grouped,
        };

        // Fallback: try turn-level lookup if call_id lookup missed
        if (!msg.reasoning_content) {
          msg.reasoning_content = sessions.getTurnReasoning(messages, msg);
        }

        messages.push(msg);
        log.debug("function_call 分组完成", { tool_calls_count: grouped.length });
      } else {
        switch (itemType) {
          case "function_call_output": {
            const callId = (item?.call_id as string) || "";
            const output = (item?.output as string) || "";
            messages.push({
              role: "tool",
              content: output,
              tool_call_id: callId,
            });
            log.debug("function_call_output 转为 tool message", { call_id: callId });
            break;
          }
          default: {
            // Regular user/assistant/developer message
            let role = (item?.role as string) || "user";
            if (role === "developer") role = "system";

            const content = valueToText(item?.content);
            const msg: ChatMessage = {
              role,
              content: content,
            };

            // For assistant messages, try to recover reasoning_content
            if (msg.role === "assistant") {
              msg.reasoning_content = sessions.getTurnReasoning(messages, msg);
            }

            messages.push(msg);
            log.debug("普通消息添加", { role, content });
          }
        }
        i++;
      }
    }
  }

  const chatReq = {
    model: req.model,
    messages,
    // Keep only `function` tools; providers like DeepSeek don't accept
    // OpenAI-proprietary built-ins (web_search, computer, file_search, …).
    tools: (req.tools || [])
      .filter((t) => (t as Record<string, unknown>)?.type === "function")
      .map(convertTool),
    temperature: req.temperature,
    max_tokens: req.max_output_tokens,
    stream: req.stream || false,
  };

  log.info("请求转换完成", {
    model: chatReq.model,
    messages_count: chatReq.messages.length,
    tools_count: chatReq.tools?.length || 0,
    stream: chatReq.stream,
  });

  return chatReq;
}

/// Responses API tool format → Chat Completions tool format.
/// Responses API (flat): {"type":"function","name":"foo","description":"...","parameters":{...}}
/// Chat Completions (nested): {"type":"function","function":{"name":"foo","description":"...","parameters":{...}}}
function convertTool(tool: unknown): unknown {
  const obj = tool as Record<string, unknown>;
  if (!obj) return tool;

  // Already in Chat Completions format if it has a "function" sub-object.
  if (obj.function) return tool;

  // Convert from Responses API flat format.
  if (obj.type === "function") {
    const func: Record<string, unknown> = {};
    if (obj.name) func.name = obj.name;
    if (obj.description) func.description = obj.description;
    if (obj.parameters) func.parameters = obj.parameters;
    if (obj.strict) func.strict = obj.strict;

    return { type: "function", function: func };
  }

  return tool;
}

/// Convert a Chat Completions response into a Responses API response.
export function fromChatResponse(
  id: string,
  model: string,
  chat: ChatResponse
): { response: ResponsesResponse; messages: ChatMessage[] } {
  log.debug("转换响应", { id, model });

  const choice: ChatChoice = chat.choices?.[0] || {
    message: { role: "assistant", content: "" },
  };

  const text = choice.message?.content || "";
  const usage: ChatUsage = chat.usage || {
    prompt_tokens: 0,
    completion_tokens: 0,
    total_tokens: 0,
  };

  const response: ResponsesResponse = {
    id,
    object: "response",
    model,
    output: [{
      type: "message",
      role: "assistant",
      content: [{
        type: "output_text",
        text,
      }],
    }],
    usage: {
      input_tokens: usage.prompt_tokens,
      output_tokens: usage.completion_tokens,
      total_tokens: usage.total_tokens,
    },
  };

  log.info("响应转换完成", {
    id,
    model,
    output_text_len: text.length,
    tool_calls_count: choice.message?.tool_calls?.length || 0,
    usage: usage,
  });

  return { response, messages: [choice.message] };
}

/// Collapse a Responses API content value (string or parts array) to plain text.
function valueToText(v: unknown): string {
  if (!v) return "";
  if (typeof v === "string") return v;
  if (Array.isArray(v)) {
    return v
      .map((p) => (p as Record<string, unknown>)?.text as string)
      .filter(Boolean)
      .join("");
  }
  return String(v);
}